Add support for VPC Endpoint Services

hintofbasil opened the pull request:
### Issue

https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/1859

<!-- Please link the GitHub issues related to this PR, if available -->

### Description

This change adds support for creating, deleting and updating VPC endpoint
services along with their permissions.  It supports the three suggested
configuration options - allowed priciples, acceptance required, and private DNS
name.

An enable annotation was also added so support enabling/disabling the feature
on a per load balancer basis.  The proposed design only supported
enabling/disabling this at the controller level.

Outputting the created VPC Endpoint Service parameters is deliberatly not
supported as there are still open questions as to the best approach to be used
and this change is already far too big and adding more to it does not seem
sensible.  This feature can be added in a follow-up change.

This change has been manually tested on a Kubernetes service.

I apologise for the large PR.  I couldn't determine suitable breakpoints to split
this change up into smaller changes.  I can do so if you can suggest the break
points.  I'm also fairly new to Golang so am very happy for any feedback as I
know the code quality isn't amazing here.

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

rifelpet committed:
Add models, managers, and synthesizers for VPC Endpoint Services

rifelpet committed:
Add the Endpoint Service sythensizer and manager to the stack deployer

rifelpet committed:
Add Endpoint Service annotations for parsing by the model

hintofbasil committed:
Add support for VPC Endpoint Services

This change adds support for creating, deleting and updating VPC endpoint
services along with their permissions.  It supports the three suggested
configuration options - allowed priciples, acceptance required, and private DNS
name.

An enable annotation was also added so support enabling/disabling the feature
on a per load balancer basis.  The proposed design only supported
enabling/disabling this at the controller level.

Outputting the created VPC Endpoint Service parameters is deliberatly not
supported as there are still open questions as to the best approach to be used
and this change is already far too big and adding more to it does not seem
sensible.  This feature can be added in a follow-up change.

maruina reviewed (commented):
Generally LGTM, just a couple of questions

hintofbasil commented on pkg/service/model_build_endpoint_service_test.go:
```diff
@@ -0,0 +1,203 @@
+				service:          tt.svc,
+			}
+			got, err := builder.buildEnabled(context.Background())
+			if tt.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.Equal(t, tt.want, got)
```
The errors returned are generally wrapped errors.  Wrapped errors include a stack trace so can't be compared an error generated in the test.
If there is a nice way to tests for equality with wrapped errors I'd be very happy to change it.

hintofbasil committed:
Remove unneeded defaultEndpointServiceManager.ReconcileTags method

This was part of the templating work done by rifelpet but the ReconcileTags
method belonging to defaultTaggingManager works for this use case.

hintofbasil commented on pkg/deploy/ec2/endpoint_service_manager.go:
```diff
@@ -0,0 +1,269 @@
+}
+
+func (m *defaultEndpointServiceManager) ReconcileTags(ctx context.Context, resID string, desiredTags map[string]string, opts ...ReconcileTagsOption) error {
+	return nil
```
This was templated by rifelpet but I ended up using the tagging manager implementation instead of duplicating the logic.  Removed it.

hintofbasil committed:
Add explanation to algorithm.DiffStringSlice

hintofbasil commented on pkg/algorithm/strings.go:
```diff
@@ -12,3 +14,25 @@ func ChunkStrings(targets []string, chunkSize int) [][]string {
 	return chunks
 }
+
+func DiffStringSlice(first, second []string) ([]*string, []*string, []*string) {
```
Added one.

kishorj commented:
@hintofbasil, thanks for your changes. We are currently finalizing the v2.4.2 patch release, I will review your changes after the patch release.

hintofbasil committed:
Revert formatting only changes

This should make the PR slightly easier to review.  Proper linting should
handle these issues in the future.

gjtempleton commented on pkg/deploy/ec2/endpoint_service_manager.go:
```diff
@@ -0,0 +1,265 @@
+	ReconcilePermissions(ctx context.Context, permissions *ec2model.VPCEndpointServicePermissions) error
+}
+
+// NewdefaultEndpointServiceManager constructs new defaultEndpointServiceManager.
```
Nit:
```suggestion
// NewDefaultEndpointServiceManager constructs new defaultEndpointServiceManager.
```

gjtempleton commented on pkg/deploy/ec2/endpoint_service_manager.go:
```diff
@@ -0,0 +1,265 @@
+	externalManagedTags []string
+}
+
+func (m *defaultEndpointServiceManager) Create(ctx context.Context, resSG *ec2model.VPCEndpointService) (ec2model.VPCEndpointServiceStatus, error) {
```
Seems there's still a few references to `SG` rather than `ES` throughout this function.
```suggestion
func (m *defaultEndpointServiceManager) Create(ctx context.Context, resES *ec2model.VPCEndpointService) (ec2model.VPCEndpointServiceStatus, error) {
```

gjtempleton commented on pkg/deploy/ec2/endpoint_service_manager.go:
```diff
@@ -0,0 +1,265 @@
+
+func (m *defaultEndpointServiceManager) Update(ctx context.Context, resES *ec2model.VPCEndpointService, sdkES networking.VPCEndpointServiceInfo) (ec2model.VPCEndpointServiceStatus, error) {
+
+	m.logger.Info("Updating", "resES", resES, "sdkES", sdkES)
```
Do we want to log this if no changes are being made?

hintofbasil replied:
In `pkg/deploy/ec2/endpoint_service_synthesizer.go` we calculate the endpoint services that need created, deleted and updates and only call the relevant function.  If no endpoint services need updated then this function is never called.  It feels safe to me.

gjtempleton commented on pkg/deploy/ec2/endpoint_service_manager.go:
```diff
@@ -0,0 +1,265 @@
+}
+
+func (m *defaultEndpointServiceManager) ReconcilePermissions(ctx context.Context, permissions *ec2model.VPCEndpointServicePermissions) error {
+	m.logger.Info("Reconciling Permissions")
```
The pattern through the rest of the codebase is lower case logging. Do we want to deviate from that here?

gjtempleton commented on docs/guide/ingress/annotations.md:
```diff
@@ -57,6 +57,10 @@ You can add annotations to kubernetes Ingress and Service objects to customize t
 |[alb.ingress.kubernetes.io/target-node-labels](#target-node-labels)|stringMap|N/A|Ingress,Service|N/A|
+|[alb.ingress.kubernetes.io/aws-load-balancer-endpoint-service-enabled](#endpoint-service-enable)|boolean|false|
+|[alb.ingress.kubernetes.io/aws-load-balancer-endpoint-service-acceptance-required](#endpoint-service-acceptance)|boolean||
+|[alb.ingress.kubernetes.io/aws-load-balancer-endpoint-service-allowed-principals](#endpoint-allowed-principles)|stringList||
```
Nit:
```suggestion
|[alb.ingress.kubernetes.io/aws-load-balancer-endpoint-service-allowed-principals](#endpoint-allowed-principals)|stringList||
```

gjtempleton reviewed (commented):
I've only got through 11 of the files, but a few nits and some style questions.

hintofbasil committed:
Spelling: Principles -> Principals

hintofbasil committed:
Fix typos

hintofbasil commented:
This has been rebased and the formatting changes above implemented.

hintofbasil commented:
I'm unable to replicate the e2e test failure.

I've added the following resources to a cluster running this branch of the controller.

```
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  name: aws-load-balancer-controller-test
spec:
  controller: ingress.k8s.aws/alb
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: aws-load-balancer-controller-test
  namespace: test
  annotations:
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/subnets: subnet-06ceeb812709c1c2b,subnet-0a3076345bb652e2f
spec:
  ingressClassName: aws-load-balancer-controller-test
  rules:
  - host: ""
    http:
      paths:
      - path: /path
        pathType: Exact
        backend:
          service:
            name: aws-load-balancer-controller-test
            port:
              name: http
---
apiVersion: v1
kind: Service
metadata:
  name: aws-load-balancer-controller-test
  namespace: test
spec:
  ports:
  - name: http
    port: 80
    targetPort: http
  selector:
    service: 'null'
  type: NodePort
```

This brings up a load balancer which successfully provisions.

I've then deleted the Ingress and the load balancer successfully deletes.  Then deleted the namespace which also successfully deletes.

From what I can see this follows the exact logic of the test.

As such I'm going to trigger a rerun to see if it was a flake.

hintofbasil commented:
/retest

seh commented:
How does this approach differ from @rifelpet's initial pass in #1948?

seh commented:
I didn't see it while reading the code, but I'm wondering whether there are more IAM actions that we'll need to allow in the IAM policy used by this controller.

hintofbasil committed:
Add PrivateLink permissions to IAM policy

hintofbasil commented:
> I didn't see it while reading the code, but I'm wondering whether there are more IAM actions that we'll need to allow in the IAM policy used by this controller.

You were correct, after adding these the tests are passing.  Thanks!

hintofbasil commented:
> How does this approach differ from @rifelpet's initial pass in https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/1948?

This is a continuation of their work.  Rifelpet provided a rough outline for the code; function definitions and structs mainly.  This adds the actual implementation to make the feature work.  It sticks pretty closely to Rifelpet's original design - and even includes their commits showing how it was built up from there.

seh commented:
> This is a continuation of their work.

Thank you for the detailed explanation.

I was going to say earlier this morning that configuring the private DNS name on the VPC endpoint service side seemed backwards to me, thinking that it would be consumers creating interface VPC endpoints that would decide on the name, but then I went back and read the AWS documentation on the feature, and see that your design matches that feature's interface: the service provider chooses a DNS name. I still find this so confusing. That's not your fault, though.

rifelpet commented on pkg/deploy/ec2/endpoint_service_manager.go:
```diff
@@ -0,0 +1,265 @@
+}
+
+func (m *defaultEndpointServiceManager) Create(ctx context.Context, resES *ec2model.VPCEndpointService) (ec2model.VPCEndpointServiceStatus, error) {
+	sgTags := m.trackingProvider.ResourceTags(resES.Stack(), resES, resES.Spec.Tags)
+	sdkTags := convertTagsToSDKTags(sgTags)
```
```suggestion
	esTags := m.trackingProvider.ResourceTags(resES.Stack(), resES, resES.Spec.Tags)
	sdkTags := convertTagsToSDKTags(esTags)
```

rifelpet commented on pkg/deploy/ec2/endpoint_service_synthesizer.go:
```diff
@@ -0,0 +1,180 @@
+	}
+
+	// We delete before we create as we can only have a single VPC end point per LB
+	for _, sdkSG := range unmatchedSDKESs {
+		if err := s.esManager.Delete(ctx, sdkSG); err != nil {
```
```suggestion
	for _, sdkES := range unmatchedSDKESs {
		if err := s.esManager.Delete(ctx, sdkES); err != nil {
```

rifelpet commented on pkg/deploy/ec2/endpoint_service_synthesizer.go:
```diff
@@ -0,0 +1,180 @@
+	sdk networking.VPCEndpointServiceInfo
+}
+
+func matchResAndSDKEndpointServices(resSGs []*ec2model.VPCEndpointService, sdkSGs []networking.VPCEndpointServiceInfo,
```
```suggestion
func matchResAndSDKEndpointServices(resESs []*ec2model.VPCEndpointService, sdkESs []networking.VPCEndpointServiceInfo,
```

rifelpet replied:
More references to 'sg' in this function that should be updated to 'es'

rifelpet commented on pkg/deploy/ec2/tagging_manager.go:
```diff
@@ -175,6 +186,48 @@ func (m *defaultTaggingManager) listSecurityGroupsWithTagFilter(ctx context.Cont
+		if err != nil {
+			return nil, err
+		}
+		for sgID, esInfo := range esInfoByIDForTagFilter {
+			esInfoByID[sgID] = esInfo
```
```suggestion
		for esID, esInfo := range esInfoByIDForTagFilter {
			esInfoByID[esID] = esInfo
```

hintofbasil committed:
Fix SG -> ES typos

hintofbasil commented:
Thanks @rifelpet, seems quite a few of these typos escaped me.
I've cleaned them all up in a single commit.

shoukoo commented:
just wondering if it's ready to be merged? would like to have this feature in the default branch.

hintofbasil committed:
Add new mocks to gen_mocks.sh

hintofbasil commented:
Looks to me like the tests timed out while still progressing.  The failed test is given 10 attempts, each taking up to 70 seconds, to call the load balancer but the test timed out after approximately two and a half minutes.

Going to try a rerun as I can't see a hard-coded timeout anywhere and I can't see how this is related to changes in the PR.

sftim commented on docs/guide/service/annotations.md:
```diff
@@ -47,6 +47,10 @@
 | [service.beta.kubernetes.io/aws-load-balancer-target-node-labels](#target-node-labels)           | stringMap               |                           |                                                        |
 | [service.beta.kubernetes.io/aws-load-balancer-attributes](#load-balancer-attributes)             | stringMap               |                           |                                                        |
 | [service.beta.kubernetes.io/aws-load-balancer-manage-backend-security-group-rules](#manage-backend-sg-rules)  | boolean    | true                      |                                                        |
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-enabled](#endpoint-service-enable)| boolean                 | false                     |                                                        |
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-acceptance-required](#endpoint-service-acceptance)| boolean |                           |                                                        |
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-allowed-principals](#endpoint-allowed-principals)| stringList |                         |                                                        |
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-private-dns-name](#endpoint-private-dns)| string            |                           |                                                        |
```
New annotations shouldn't really contain “beta” in their names (see Kubernetes API design docs).

hintofbasil replied:
There have been updated to be alpha instead of beta.

johngmyers commented on docs/guide/service/annotations.md:
```diff
@@ -47,6 +47,10 @@
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-enabled](#endpoint-service-enable)| boolean                 | false                     |                                                        |
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-acceptance-required](#endpoint-service-acceptance)| boolean |                           |                                                        |
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-allowed-principals](#endpoint-allowed-principals)| stringList |                         |                                                        |
+| [service.beta.kubernetes.io/aws-load-balancer-endpoint-service-private-dns-name](#endpoint-private-dns)| string            |                           |                                                        |
```
What's the MergeBehavior of these new annotations?

hintofbasil replied:
I'm not sure I understand the question.  Could you clarify it?

gosunilgo commented:
@hintofbasil  - Do you have time/bandwidth to rebase this week?   Need help?   (Eager to try this)   Thanks for merge!

hintofbasil commented:
> @hintofbasil - Do you have time/bandwidth to rebase this week? Need help? (Eager to try this) Thanks for merge!

Apologies.  I didn't realise it had been so long since I looked at this.

I will try and put some time aside to rebase it this week.


We've been running this branch in production for some time now successfully but it would be great to get it merged upstream so we can get all the recent updates too.

hintofbasil committed:
Merge branch 'main' of github.com:kubernetes-sigs/aws-load-balancer-controller into endpoint-service

hintofbasil committed:
Use gomock for MockProvider

Whenr this was originally written Provider (or DefaultProvider at the time) was
not an interface only a struct so was not mockable.  Now that it's an interface
we can replace the bad solution with a proper mock implementation.

hintofbasil commented:
Going to try rerunning the tests in case the issue was a flake as it has that feel about it.

hintofbasil commented:
/retest

cgagnon-crakmn commented:
@hintofbasil I am following your PR closely and eager to try this. Hope you have time to finish this soon!

hintofbasil committed:
Update VPCES annotations to be alpha

These annotations are new and should be introduced under and alpha
flag instead of the beta one currently used for everything else.

There is no default functionality for changing the version of an
annotation but we can explicitly add the alpha annotation
and assume users will follow the docs and use the alpha annotation
despite the beta annotation still actually working.

hintofbasil committed:
Merge branch 'main' of github.com:kubernetes-sigs/aws-load-balancer-controller into endpoint-service

hintofbasil commented:
I think that all the issues raised so far in this PR have now been addressed.  If I have missed one then please raise it again.

The PR as a whole is ready for review again.

I will make sure to fix any merge conflicts more promptly and try my best to keep this in a review-ready state from now on.

trxa commented:
@hintofbasil I tested your PR and it creates the VPCES but deletion does not work. The LB controller is stuck with error:

`error":"ResourceInUse: Load balancer 'arn:...' cannot be deleted because it is currently associated with another service\n\tstatus code: 400`

The ES needs to be deleted first. Before that, probably all endpoints need to be rejected.

I also cannot see the endpoint service name. As it cannot be specified it would be better to attach it to the Service CR when the ES is created. Maybe as message in a condition.

hintofbasil committed:
Merge branch 'main' of github.com:kubernetes-sigs/aws-load-balancer-controller into endpoint-service

hintofbasil commented:
>  tested your PR and it creates the VPCES but deletion does not work. The LB controller is stuck with error:
> 
> `error":"ResourceInUse: Load balancer 'arn:...' cannot be deleted because it is currently associated with another service\n\tstatus code: 400`
> 
> The ES needs to be deleted first. Before that, probably all endpoints need to be rejected.

If you have connected an endpoint to your endpoint service then there is little I can do here as AWS doesn't let you delete an endpoint service with connected endpoints.  As the endpoints are unknown to the controller these can not be automatically cleaned up.  What would you imagine happening in this case?

If I've misunderstood the problem then let me know.

> I also cannot see the endpoint service name. As it cannot be specified it would be better to attach it to the Service resource when the ES is created. Maybe as message in a condition.

I agree it would be useful to expose this somewhere.  But to avoid the PR ballooning even more I would suggest this be added in the follow-up PR.

trxa commented:
You are right. Rejecting the connection is not sufficient to delete the service. Didn't realise it. Anyway the removal attempt for the ES should be done first and the operator should wait until it is possible and proceed then with the load balancer. It allows proper cleanup of landscapes without involving the AWS API.

hintofbasil committed:
Handle delete VPCEndpoint failures

These failures are in the returned object not the returned
error.

hintofbasil committed:
Move VPCES creation and update to post synthesize

These need to happen after the load balancer has synthesized otherwise
there may be no load balancer for them to attach to.

We also need to synthesize the VPC endpoint service before the load
balancers as VPC endpoint services will need deleted before load
balancers are deleted.

hintofbasil commented:
> You are right. Rejecting the connection is not sufficient to delete the service. Didn't realise it. Anyway the removal attempt for the ES should be done first and the operator should wait until it is possible and proceed then with the load balancer. It allows proper cleanup of landscapes without involving the AWS API.

You're correct.  It turns out there was a bug which wasn't catching the failure to delete the VPC endpoint service.  But had the bug not been there it wouldn't have worked anyway as the ordering of creations / deletions was incorrect.

I've added some commits to fix this.  It should now delete the VPC endpoint service before attempting to delete the load balancer.
It will also create the load balancer before attempting to create the VPC endpoint service.  It won't wait for the load balancer to be ready before continuing though so there may be a small flurry of errors during the interim time.  But they are harmless and the exponential backoff of the controller deals with them well.

Feel free to give it another shot and see if the issue is resolved.

trxa commented:
I tried it again and now it's deleting.

hintofbasil committed:
Merge branch 'main' of github.com:kubernetes-sigs/aws-load-balancer-controller into endpoint-service

tculp commented:
Any update on this?

hintofbasil commented:
> Any update on this?

The code is up to date on my end.  Just waiting for a review from kishorj and/or M00nf1sh.  I presume it being added to the v2.6.0 roadmap means they are aware it still needs review.

dims commented:
/assign @oliviassss

oliviassss commented:
Thanks for your contribution, and sorry for the delay. I will discuss and review this PR internally with our approvers to gather opinions.

oliviassss commented:
Sorry for the delay, we discussed this feature internally, it's a major feature that needs a minor release, and also we need to do a security review with it. So we aim at shipping it with v2.8.0, Q2 2024

aaroniscode commented:
/remove-lifecycle stale

prianna commented:
Hey @oliviassss AWS has released 2.8.1 of the LB controller and this PR still doesn't seem to have been merged. I don't see support for managing VPC private endpoints natively in the controller's release notes. Any idea what the blocker for this is? We'd love an update, as we have several major architectural changes that depend on this at Salesforce.

oliviassss commented:
/assign @M00nF1sh

hintofbasil committed:
Merge branch 'main' of https://github.com/kubernetes-sigs/aws-load-balancer-controller into endpoint-service

m17kea commented:
Any update on this?

bsod90 commented:
It would be so awesome if this one got merged. We're considering implementing an alternative solution, but doing it through the load balancer controller would be so much cleaner.

sudheer-gutta commented:
We already have 2.10.0 version of load-balancer-controller and this PR is not part of it. We are trying to design a solution using multiple VPC endpoint services and having native support is helpful. Can you please share the plan on when this PR will make it into a release?

vazuev commented:
Hello @oliviassss @kishorj @M00nF1sh !
This feature would be really liked in ALB-controller solution!

z0rc commented:
/remove-lifecycle stale

vincentmrg commented:
Hello @oliviassss @kishorj @M00nF1sh 
Would it be possible, please, to share what is planned on your side regarding the integration of this feature in a future release?

m17kea commented:
/remove-lifecycle stale
